### PR TITLE
UPDATE: DataMapper Implementation

### PR DESCRIPTION
**Pull request by SorenNorgeA**
New GPCRdb tools for users to create data visualization of their own data. The DataMapper provides five different visualizations: GPCRome, Tree, Cluster, List, and Heatmap. All plots provides various 'on-the-fly' customization features for the user to generate their desired visualization of their data.

### CODE REVIEW

File: mapper/views.py [hunk 1]
```diff
@@ -0,0 +1,1278 @@
...
+                                  ('children',data_b2['children'])])
+        class_c_dict = OrderedDict([('name', 'Class C (Glutamate)'),
+                                  ('value', 0),
+                                  ('color', 'Purple'),
+                                  ('children',data_c['children'])])
+        class_f_dict = OrderedDict([('name', 'Class F (Frizzled)'),
+                                  ('value', 0),
+                                  ('color', 'Grey'),
+                                  ('children',data_f['children'])])
+        class_t2_dict = OrderedDict([('name', 'Class T (Taste 2)'),
```
Comment (pszgaspar) [inline mapper/views.py:277]: please change to new class name: Class T2 (Taste 2)

File: mapper/views.py [hunk 2]
```diff
@@ -0,0 +1,1127 @@
...
+                                try:
+                                    header_list = [cell.value for cell in worksheet[1]]
+                                except:
+                                    return render(request, self.template_name, {'upload_status': 'Failed','Error_message': "Corrupted excel, headers not inline with the template file."})
+
+                                # If first sheet is receptor with correct headers #
+                                if sheet_name == 'Tree':
+
+                                    header = next(worksheet.iter_rows(min_row=1, max_row=1, values_only=True))
+                                    # Labels = next(worksheet.iter_rows(min_row=2, max_row=1, values_only=True))
```
Comment (pszgaspar) [inline mapper/views.py:557]: please clean up

File: mapper/views.py [hunk 3]
```diff
@@ -0,0 +1,1127 @@
...
+
+
+                    else:
+                        return render(request, self.template_name, {'upload_status': 'Failed','Error_message': "Unable to load excel file, might be corrupted or not inline with the template file."})
+                    # Needs to send a response if everything is handled #
+                    #return render(request, self.template_name, {'upload_status': 'Success'})
+
+            else:
+                # Return a 405 Method Not Allowed response if not a POST request
+                #return HttpResponseNotAllowed(['POST'])
```
Comment (pszgaspar) [inline mapper/views.py:1032]: please clean up

File: mapper/views.py [hunk 4]
```diff
@@ -0,0 +1,1127 @@
...
+                return render(request, self.template_name, {'upload_status': 'Failed','Error_message': "Not a valid excel file. Please try and use the template excel file."})
+
+# def LandingPage(request):
```
Comment (pszgaspar) [inline mapper/views.py:1035]: this function is not needed I guess? Please remove

File: static/home/images/dart_plot.png
Comment (pszgaspar) [inline static/home/images/dart_plot.png:1]: is this replaced by the gpcrome_example? If yes, please remove

File: static/home/js/datamapper.js [hunk 5]
```diff
@@ -0,0 +1,3548 @@
...
+        }
+
+        colorScales[category] = gradientId;
+    });
+
+    // var barWidth = width / existingCategories.length - 20; // Adjust the width of each bar
+    var barWidth = 120;
+    var barHeight = 15;
+    var spacing = 50; // Horizontal spacing between bars
+    // console.log(barWidth);
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:483]: please remove

File: static/home/js/datamapper.js [hunk 6]
```diff
@@ -0,0 +1,3548 @@
...
+    // Check the visibility of each layer
+    const Layer1_isChecked = d3.select(`#toggle-layer-1`).property('checked');
+    const Layer2_isChecked = d3.select(`#toggle-layer-2`).property('checked');
+    const Layer3_isChecked = d3.select(`#toggle-layer-3`).property('checked');
+
+    let Checklist = [Layer1_isChecked, Layer2_isChecked, Layer3_isChecked];
+
+    // Indentation
+
+    // Data sorter depending on layers being shown:
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:905]: couple of commented functions below, will we need these?

File: static/home/js/datamapper.js [hunk 7]
```diff
@@ -0,0 +1,3548 @@
...
+        for (const [entity, char] of Object.entries(htmlEntities)) {
+            transformedLabel = transformedLabel.replace(new RegExp(entity, 'g'), char);
+        }
+        // Remove <i> and </i> tags from the label
+        transformedLabel = transformedLabel.replace(/<\/?i>/g, '');
+        // remove receptor and adrenoceptor
+        transformedLabel = transformedLabel.replace("-adrenoceptor", '').replace(" receptor", '');
+        // Clear existing content in textElement if needed
+        textElement.text('');
+        // textElement.text(transformedLabel);
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:2247]: please clean up

File: static/home/js/datamapper.js [hunk 8]
```diff
@@ -0,0 +1,3548 @@
...
+      .attr('y', 50)
+      .style("font-size", `${data_fontsize}px`)
+      .style("font-family", "sans-serif")
+      .style("text-anchor", "end")
+      .text(highest_value);
+
+    // Rerender height of plot as the last thing
+    let label_length_final;
+    if (rotation == 90) {
+        // label_length_final = Math.ceil(longestLabel*0.8 + 13.25*label_fontsize)
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:2569]: please clean

File: static/home/js/datamapper.js [hunk 9]
```diff
@@ -0,0 +1,3548 @@
...
+            // If it's a header, set the rotation to 0, otherwise calculate the outward-facing rotation
+            let rotation;
+            if (isInHeaderList) {
+                rotation = 0;
+            } else if (text_value === "Classless") {
+                rotation = 0;
+            } else {
+                rotation = -(angle * 180 / Math.PI);
+            }
+            // const rotation = (isInHeaderList || "Classless") ? 0 : -(angle * 180 / Math.PI);
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:2991]: clean

File: static/home/js/datamapper.js [hunk 10]
```diff
@@ -0,0 +1,3548 @@
...
+            // Apply the rotation and positioning
+            return `rotate(${pos.rotation + rotation}, ${pos.x}, ${pos.y})`;
+        })
+        .html(d => GPCRome_formatTextWithHTML(d, Family_list, level))
+        .style("font-size", d => Header_list.includes(d) ? "26px" : "9px")
+        .style("font-family", "Palatino")
+        .attr("font-weight", d => Header_list.includes(d) || Family_list.includes(d) ? "950" : "normal")
+        .style("fill", d => Header_list.includes(d) ? "Black" : "black");
+
+            // .style("background",'black')
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:3056]: clean

File: static/home/js/datamapper.js [hunk 11]
```diff
@@ -0,0 +1,3548 @@
...
+                const value = fill_data[d.data]?.Value1;
+                if (value === "Yes" || value === "No" || !isNaN(value)) return "black";
+                return "none";  // Remove the stroke if the value is ""
+            })
+            .style("stroke-width", (d) => {
+                const value = fill_data[d.data]?.Value1;
+                return value === "" ? 0 : 0.5;  // Set stroke-width to 0 if the value is an empty string
+            });
+    }
+    // draw a clock
```
Comment (pszgaspar) [inline static/home/js/datamapper.js:3251]: spiral maybe for later but the clock? Can we remove?

**Review by pszgaspar — CHANGES_REQUESTED** [review]
Very nice guys! I just found a couple of cleanup related things

File: mapper/views.py (hunk 1 above)
Comment (Jimmy-A-Caroli) [reply mapper/views.py]: Fixed this!

File: static/home/js/datamapper.js (hunk 5 above)
Comment (Jimmy-A-Caroli) [reply static/home/js/datamapper.js:483]: Removed with latest commit

File: static/home/js/datamapper.js (hunk 6 above)
Comment (Jimmy-A-Caroli) [reply static/home/js/datamapper.js:905]: Not sure, but I think these are still WIP

File: static/home/images/dart_plot.png
Comment (Jimmy-A-Caroli) [reply static/home/images/dart_plot.png:1]: Removed with latest commit

**Review by pszgaspar — APPROVED** [review]

**Pull request merged**